File: lib/core/services/chat_service.dart
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:hilla_ride/core/models/app_models.dart';
import 'package:hilla_ride/core/models/chat_models.dart';
import 'package:uuid/uuid.dart';

class ChatService {
  ChatService({FirebaseFirestore? firestore})
      : _firestore = firestore ?? FirebaseFirestore.instance;

  final FirebaseFirestore _firestore;
  final _uuid = const Uuid();

  Stream<List<RideMessage>> watchRideMessages(String rideId) {
    return _firestore
        .collection('rides')
        .doc(rideId)
        .collection('messages')
        .orderBy('createdAt', descending: false)
        .snapshots()
        .map((snapshot) {
      return snapshot.docs
          .map((doc) => RideMessage.fromMap(doc.id, doc.data()))
          .toList();
    });
  }

  Future<void> sendRideMessage({
    required String rideId,
    required String senderId,
    required UserRole senderRole,
    required String senderName,
    required String text,
  }) async {
    final trimmed = text.trim();
    if (trimmed.isEmpty) return;

    await _firestore
        .collection('rides')
        .doc(rideId)
        .collection('messages')
        .doc(_uuid.v4())
        .set({
      'senderId': senderId,
      'senderRole': senderRole.name,
      'senderName': senderName,
      'text': trimmed,
      'type': RideMessageType.text.value,
      'createdAt': FieldValue.serverTimestamp(),
    });
  }

  Future<void> sendRideVoiceMessage({
    required String rideId,
    required String senderId,
    required UserRole senderRole,
    required String senderName,
    required String voiceUrl,
    required int voiceDurationMs,
  }) async {
    if (voiceUrl.trim().isEmpty) return;

    await _firestore
        .collection('rides')
        .doc(rideId)
        .collection('messages')
        .doc(_uuid.v4())
        .set({
      'senderId': senderId,
      'senderRole': senderRole.name,
      'senderName': senderName,
      'text': '',
      'type': RideMessageType.voice.value,
      'voiceUrl': voiceUrl,
      'voiceDurationMs': voiceDurationMs,
      'createdAt': FieldValue.serverTimestamp(),
    });
  }
}

File: hilla_ride_ios/HelloTukTuk/Core/Services/ChatService.swift
import FirebaseFirestore
import Foundation

struct RideMessage: Identifiable, Equatable {
    let id: String
    let senderId: String
    let senderRole: UserRole
    let senderName: String
    let text: String
    let createdAt: Date?

    init?(documentID: String, data: [String: Any]) {
        id = documentID
        senderId = data["senderId"] as? String ?? ""
        senderName = data["senderName"] as? String ?? ""
        text = data["text"] as? String ?? ""
        senderRole = UserRole.fromFirestore(data["senderRole"] as? String)
        if let timestamp = data["createdAt"] as? Timestamp {
            createdAt = timestamp.dateValue()
        } else {
            createdAt = nil
        }
    }
}

final class ChatService {
    private let firestore = Firestore.firestore()

    func watchRideMessages(rideId: String) -> AsyncStream<[RideMessage]> {
        AsyncStream { continuation in
            let listener = firestore.collection("rides").document(rideId)
                .collection("messages")
                .order(by: "createdAt", descending: false)
                .addSnapshotListener { snapshot, _ in
                    let messages = snapshot?.documents.compactMap {
                        RideMessage(documentID: $0.documentID, data: $0.data())
                    } ?? []
                    continuation.yield(messages)
                }
            continuation.onTermination = { _ in listener.remove() }
        }
    }

    func sendRideMessage(
        rideId: String,
        senderId: String,
        senderRole: UserRole,
        senderName: String,
        text: String
    ) async throws {
        let trimmed = text.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return }

        try await firestore.collection("rides").document(rideId)
            .collection("messages")
            .document(UUID().uuidString)
            .setData([
                "senderId": senderId,
                "senderRole": senderRole.rawValue,
                "senderName": senderName,
                "text": trimmed,
                "type": "text",
                "createdAt": FieldValue.serverTimestamp()
            ])
    }
}
